lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -276,536 +344,505 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
 }
 
 // =====================================================================================================================
-// Generates the type for the new entry-point based on already-collected info in LLVM context.
+// Generates the type for the new entry-point based on already-collected info.
+// This is what decides what SGPRs and VGPRs are passed to the shader at wave dispatch:
+//
+// * (For a GFX9+ merged shader or NGG primitive shader, the 8 system SGPRs at the start are not accounted for here.)
+// * The "user data" SGPRs, up to 32 (GFX9+ non-compute shader) or 16 (compute shader or <=GFX8). Many of the values
+//   here are pointers, but are passed as a single 32-bit register and then expanded to 64-bit in the shader code:
+//   - The "global information table", containing various descriptors such as the inter-shader rings
+//   - The "per-shader table", which is added here but appears to be unused
+//   - The streamout table if needed
+//   - Nodes from the root user data layout, including pointers to descriptor sets. In a compute shader, these
+//     must have the same layout as the root user data layout, even if that would mean leaving gaps in register
+//     usage, and are limited to 10 SGPRs s2-s11
+//   - For a compute shader, the spill table pointer if needed, which must be in s12
```

Comment:
The spill table pointer should be setup in all shader types that need it, no?

From skimming PAL code, it also seems to be that having the spill table in s12 is not a strict requirement of PAL, though it _is_ a good idea for LLPC to always use the same SGPR for it (it can avoid extraneous PM4-based register writes in some cases).

Similarly, it looks like the spill table pointer can be in any user data SGPR for graphics pipelines, though of course it's natural to just put it into the last one...

Reply:
You're probably right, but this change deliberately didn't change anything from the old code so I could test it by just comparing generated code and metadata.

Reply:
For graphics, I think this code does not necessarily put the spill table in the last sgpr. Suppose there are two compact descriptors left when we have two sgprs left. It would put the first descriptor in s30-s31, then discover it needs to spill the second one, then spill the first one so it has room for the spill table. The spill table then gets into s30.

Reply:
Good point about the graphics spill table pointer.

In the unlikely event that an application changes the pipeline _without_ changing any user data registers, always putting the spill table pointer into s15 / s31 would have the advantage that PAL wouldn't have to update it. That seems exceedingly unlikely though, especially for graphics shaders, and in any case the strategy of not changing behavior in this PR is a good one.

Reply:
I still think it's a good idea to slightly edit the comment to reflect our best understanding of the situation, though ;)